File: src/Wc3InputLua/Selection.lua
--=========
-- Include
--=========

---@type Wc3Utils
local Utils = LibManager.getDepency('Wc3Utils')
local ActionList = Utils.ActionList or error('')

--========
-- Module
--========

---@class InputSelection
local Selection = {}

local lock = false
local group = {}
local actions = ActionList.new()

---@param group table<integer, Unit>
---@return table<integer, Unit>
local function clearDead(group)
    local new_group = {}
    for i = 1, #group do
        if group[i]:getHealth() > 0.5 then
            table.insert(new_group, group[i])
        end
    end
    return new_group
end

local function onSelection()
    local pl = GetTriggerPlayer()
    local u = GetTriggerUnit()
    local gr = group[pl]

    if not u then return end

    local found = -1
    for i = 1, #gr do
        if gr[i] == u then
            found = i
            break
        end
    end

    if lock then
        if found < 0 and pl == GetLocalPlayer() then
            SelectUnit(u, false)
        end
    else
        if found < 0 then
            table.insert(gr, u)
            actions:run(gr, pl)
        end
    end
end

local function onDeselection()
    local pl = GetTriggerPlayer()
    local u = GetTriggerUnit()
    local gr = group[pl]

    if not u then return end

    local found = -1
    for i = 1, #gr do
        if gr[i] == u then
            found = i
            break
        end
    end

    if lock then
        if found < 0 and pl == GetLocalPlayer() then
            SelectUnit(u, true)
        end
    else
        if found > 0 then
            table.remove(gr, found)
            actions:run(gr, pl)
        end
    end
end

---@param flag boolean
---@param pl player | nil
function Selection.lock(flag, pl)
    pl = pl or GetLocalPlayer()
    if pl ~= GetLocalPlayer() then
        return
    end

    lock = flag
    EnableSelect(not flag, true)
    EnableDragSelect(not flag, true)
    ClearSelection()

    local gr = group[pl]
    for i = 1, #gr do
        SelectUnit(gr[i], true)
    end
end

---@alias InputSelectionCallback fun(group:table<number, unit>, pl:player)

---@param callback InputSelectionCallback
---@return Action
function Selection.addAction(callback)
    return actions:add(callback)
end

---@param action Action
---@return boolean
function Selection.removeAction(action)
    return actions:remove(action)
end

if IsGame() then
    local trigger_select = CreateTrigger()
    TriggerAddAction(trigger_select, onSelection)

    local trigger_deselect = CreateTrigger()
    TriggerAddAction(trigger_deselect, onDeselection)

    for i = 0, bj_MAX_PLAYER_SLOTS - 1 do
        local pl = Player(i)
        if GetPlayerController(pl) == MAP_CONTROL_USER and
           GetPlayerSlotState(pl) == PLAYER_SLOT_STATE_PLAYING then
            group[pl] = {}
            TriggerRegisterPlayerUnitEvent(trigger_select, pl, EVENT_PLAYER_UNIT_SELECTED)
            TriggerRegisterPlayerUnitEvent(trigger_deselect, pl, EVENT_PLAYER_UNIT_DESELECTED)
        end
    end
end

return Selection

File: src/Wc3Input/Selection.ts
import { Action, ActionList } from "../Wc3Utils/index"

export namespace Selection {
    export function lock(flag: boolean, pl: jplayer|undefined){
        pl = pl ? pl : GetLocalPlayer()
        if (pl != GetLocalPlayer()){return}

        _lock = flag
        EnableSelect(!flag, true)
        EnableDragSelect(!flag, true)
        ClearSelection()

        let gr = _groups.get(pl)
        if (!gr){return}
        for (let u of gr){
            SelectUnit(u, true)
        }
    }

    export function addAction(callback: (this: void, pl: jplayer, gr: junit[])=>void){
        return _actions.add(callback)
    }

    export function removeAction(action: Action<[jplayer, junit[]], void>){
        return _actions.remove(action)
    }

    let _actions = new ActionList<[jplayer, junit[]]>()

    function _runActios(event: 'SELECT' | 'DESELECT'){
        let pl = GetTriggerPlayer()
        let u = GetTriggerUnit()
        let gr = _groups.get(pl)
        if (!gr || !u){return}

        let found = gr.findIndex((cur: junit, i:number, gr: junit[]): boolean => {return cur == u})

        if (_lock){
            if (found < 0 && pl == GetLocalPlayer()){
                SelectUnit(u, event == 'DESELECT')
            }
        } else {
            if (event == 'SELECT' && found < 0){
                gr.push(u)
                _actions.run(pl, Object.assign([], gr))
            } else if (event == 'DESELECT' && found >= 0){
                gr.splice(found, 1)
                _actions.run(pl, Object.assign([], gr))
            }
        }
    }

    let _lock = false
    let _groups = new Map<jplayer, junit[]>()
    let _trigger_select: jtrigger | undefined
    let _trigger_deselect: jtrigger | undefined
    if (IsGame()){
        _trigger_select = CreateTrigger()
        _trigger_deselect = CreateTrigger()
        TriggerAddAction(_trigger_select, (():void => {_runActios('SELECT')}))
        TriggerAddAction(_trigger_deselect, (():void => {_runActios('DESELECT')}))

        for (let i = 0; i < bj_MAX_PLAYER_SLOTS; i++){
            let pl = Player(i)
            if (GetPlayerController(pl) == MAP_CONTROL_USER &&
                GetPlayerSlotState(pl) == PLAYER_SLOT_STATE_PLAYING){

                _groups.set(pl, [])
                TriggerRegisterPlayerUnitEvent(_trigger_select, pl, EVENT_PLAYER_UNIT_SELECTED)
                TriggerRegisterPlayerUnitEvent(_trigger_deselect, pl, EVENT_PLAYER_UNIT_DESELECTED)
            } 
        }
    }
}